### PR TITLE
Add optional exercises for section 1 and 2

### LINKED ISSUE #258: Add optional exercises at the end of each section
### As a fast learner I want to have optional exercises/assignments at the end of each section.

When teaching this lesson in person (#159), we notice that some participants already finish a section when other participants are only halfway. We want everyone to start on a new section at the same time, so that we can also introduce some concepts for that section to everyone at the same time. It would be nice to have optional exercises/assignments at the end of each section for the fast learners.

### PR DESCRIPTION
**Pull request by svenvanderburg**
Fixes #258 for section 1 and 2.

This adds a new episode at the end of the section with optional exercises. We could also choose to have 1 episode as extra material with optional exercises for all episodes.

Current status: section 1 and 2 done. This can be reviewed now or wait for all sections to have optional exercises, I think it would be nice to review now to make sure this is in line with thoughts the other contributors

### CODE REVIEW

File: _episodes/17-section1-optional-exercises.md
Comment (bielsnohr) [inline _episodes/17-section1-optional-exercises.md:43]: Given change to Anaconda licensing that make it basically unusable in research, we should probably additionally link to [miniforge](https://github.com/conda-forge/miniforge) saying that it is the tool they should install. The documentation for conda itself is still useful since that basically stays the same.

File: _episodes/17-section1-optional-exercises.md
Comment (bielsnohr) [inline _episodes/17-section1-optional-exercises.md:47]: Nice, this is a great concrete exercise :+1:

File: _episodes/25-section2-optional-exercises.md
Comment (bielsnohr) [inline _episodes/25-section2-optional-exercises.md:15]:
```suggestion
The exercises have an explorative nature, so feel free to go off in any direction that interests you.
You will be looking at some tools that either complement or are alternatives to those already introduced.
Even if you find something that you really like,
we still recommend sticking with the tools that were introduced prior to this episode when you move onto other sections of the course.
```

File: _episodes/25-section2-optional-exercises.md
Comment (bielsnohr) [inline _episodes/25-section2-optional-exercises.md:19]: As with the optional exercise in Section 1, probably good to give just a bit more guidance, e.g. trying adding some unit tests with pytest to an existing project, add CI to an existing project that runs these tests, think of some edge cases that you might not have tested, etc.

File: _episodes/25-section2-optional-exercises.md
Comment (bielsnohr) [inline _episodes/25-section2-optional-exercises.md:25]:
```suggestion
> [branch coverage](https://about.codecov.io/blog/line-or-branch-coverage-which-type-is-right-for-you/)
> as opposed to just line coverage. 
```

**Review by bielsnohr — CHANGES_REQUESTED** [review]
Thanks for the really nice contributions @svenvanderburg ! I think these nicely complement the course material and provide a great way to handle learners with a range of experience.

I have left some comments to improve the exercises you has suggested and it would be nice to incorporate these changes if possible and you have time.

An overall request I would make is that is would be nice to link to these additional exercises at the end of each relevant episode. For example, we could link to the "Try out different Integrated Development Environments" additional exercise at the end of  the "Integrated Software Development Environments". This would then accommodate the instructor-led version of the course where we synchronise at the end of each episode rather than at the end of a section (which I suspect is your experience of delivering the course?). Let me know if that isn't clear.

File: _episodes/17-section1-optional-exercises.md
Comment (bielsnohr) [inline _episodes/17-section1-optional-exercises.md:2]: Title case to be consistent with the rest of the lesson.

```suggestion
title: "Optional Exercises for Section 1"
```

File: _episodes/17-section1-optional-exercises.md
Comment (svenvanderburg) [reply _episodes/17-section1-optional-exercises.md:26]: Good point! I summarized this in this note: https://github.com/carpentries-incubator/python-intermediate-development/pull/375/commits/847e674d42839550922e6e76995a924f3d2b2ced and use this as an opportunity for the learner to reflect on what they like better.

File: _episodes/17-section1-optional-exercises.md
Comment (svenvanderburg) [reply _episodes/17-section1-optional-exercises.md]: Good point! Attempt to resolve with: 19329a53dfe53810

File: _episodes/17-section1-optional-exercises.md
Comment (svenvanderburg) [reply _episodes/17-section1-optional-exercises.md:56]: Thanks! 🙏

File: _episodes/25-section2-optional-exercises.md
Comment (svenvanderburg) [reply _episodes/25-section2-optional-exercises.md]: Yes, good one! Added with 1c007f18095c24d318ac0c8

File: _episodes/17-section1-optional-exercises.md
Comment (svenvanderburg) [reply _episodes/17-section1-optional-exercises.md:33]: Added with 44e57a1fb656300bd2ed3180

File: _episodes/17-section1-optional-exercises.md
Comment (svenvanderburg) [reply _episodes/17-section1-optional-exercises.md]: Ha, love the slogan of fish!

These are great additions! I added them here: d85e848f0a05

**Comment by svenvanderburg** [comment]
> Thanks for the really nice contributions @svenvanderburg ! I think these nicely complement the course material and provide a great way to handle learners with a range of experience.
> 
> I have left some comments to improve the exercises you has suggested and it would be nice to incorporate these changes if possible and you have time.
> 
> An overall request I would make is that is would be nice to link to these additional exercises at the end of each relevant episode. For example, we could link to the "Try out different Integrated Development Environments" additional exercise at the end of the "Integrated Software Development Environments". This would then accommodate the instructor-led version of the course where we synchronise at the end of each episode rather than at the end of a section (which I suspect is your experience of delivering the course?). Let me know if that isn't clear.

Thanks for reviewing @bielsnohr! That's a nice suggestion to put links in each episode. We actually synchronize at the end of each section, or in fact at the beginning of the next section. Funny, because that also works well. I am not sure if this is unclear or that we just did not fully read the instructor notes ;)

It would look like this:  ecabd8c9f609cf74ac32d154

An alternative would be to put the optional exercise in each episode. If this is how you usually teach it that sounds more logical to me. (Although then the comment on explorative nature of the optional exercises has to be repeated a lot or be put in the introduction).

**Comment by svenvanderburg** [comment]
@bielsnohr I now added a reference to the optional exercises from the relevant episodes.

**Comment by bielsnohr** [comment]
Thanks for the further follow up.

> That's a nice suggestion to put links in each episode. We actually synchronize at the end of each section, or in fact at the beginning of the next section. Funny, because that also works well. I am not sure if this is unclear or that we just did not fully read the instructor notes ;)

We deliberately leave this up to instructors to decide. As we note in [Course Delivery](https://carpentries-incubator.github.io/python-intermediate-development/guide/index.html#course-delivery), there are a few different approaches to take. If you are doing the more "group self-study" option, then yes, it probably makes sense to synchronise at the end of a session/day. On the other hand, using the slides and mixing in a bit of instructor-led teaching means that more frequent synchronisation at the end of each episode can also work.

> An alternative would be to put the optional exercise in each episode. If this is how you usually teach it that sounds more logical to me. (Although then the comment on explorative nature of the optional exercises has to be repeated a lot or be put in the introduction).

I think it is fine to have them as a separate episode for now. That aligns better with the independent study mode of teaching because you want learners to get through all of the core content first, and then come to the optional bits at the end of the session. On the other hand, if you are synchronising at the end of the episode, more experienced learners will be able to click the link and navigate to the optional exercises. That extra click isn't a big deal in my opinion.

Last thing from my side: could you enable maintainer contributions like described in [this GitHub page](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests)? There is a batch of formatting changes and an incorrect link I would like to submit as one commit rather than doing the "make a suggestion" feature of the PR interface. Thanks!

**Comment by svenvanderburg** [comment]
@bielsnohr cool, we will leave them as a separate episode!

In pull requests coming from an organisation (in this case from esciencecenter-digital-skills) it is not possible to allow maintainers to commit to the branch. See https://github.com/orgs/community/discussions/5634, it is  apparently a reported issue for years. As a workaround we could merge this PR into a branch on the upstream repo, then you can add your changes. By now I am actually a collaborator, so I could also commit to that new branch if needed.

**Comment by bielsnohr** [comment]
> In pull requests coming from an organisation (in this case from esciencecenter-digital-skills) it is not possible to allow maintainers to commit to the branch. See https://github.com/orgs/community/discussions/5634, it is apparently a reported issue for years. As a workaround we could merge this PR into a branch on the upstream repo, then you can add your changes. By now I am actually a collaborator, so I could also commit to that new branch if needed.

Ah, okay. Thanks for tracking that down. I was wondering why it didn't happen automatically :wink: I like your suggestion of merging this into a branch on this upstream repo. Then I can push my changes to that branch, and we then merge that into the `gh-pages` branch to become part of the material.

**Review by bielsnohr — APPROVED** [review]

**Pull request merged**
